### PR TITLE
Add neigh_rdns to BFDEvent

### LINKED ISSUE #199: BFD: missing attribute neigh_rdns
The attribute `neigh_rdns` should include "reverse dns" information of the BFD neighbor. (reverse dns of the address in the bfdaddr field i think)

the zino1 server is not consistent on this field on all types containing ipv4/ipv6 neighbor info, but this could certanly also be done on eg. BGP alarms.

This field is not critical for eg. curitz to function at this state

### PR DESCRIPTION
**Pull request by stveit**
Fixes #199

### CODE REVIEW

File: src/zino/utils.py [hunk 1]
```diff
@@ -25,3 +27,10 @@ def _parse_hexa_string_ip(ip: str) -> IPAddress:
 def _parse_colon_separated_ip(ip: str) -> IPAddress:
     """Parses IP addresses formatted with a colon symbol separating every octet, e.g. 7F:00:00:01"""
     return ip_address(bytes(OctetString(hexValue=ip.replace(":", ""))))
+
+
+def reverse_dns(ip: str) -> Optional[str]:
+    try:
+        return socket.gethostbyaddr(str(ip))[0]
+    except socket.herror:
+        return None
```
Comment (lunkwill42) [inline src/zino/utils.py:36]: This is a blocking socket operation and does not belong in an async event loop at all.

If we need to do async DNS requests, we need to look into other libraries, maybe something like `aiodnsresolver`.

**Review by lunkwill42 — CHANGES_REQUESTED** [review]
The blocking socket is a show-stopper for this.

Otherwise, I'd ask about the original issue as well. I could not find `neigh_rdns` in the original Zino 1 codebase, but rather it seems the event attribute is named `Neigh-rDNS` (verbatim).  Does this matter, @runborg ?

File: src/zino/utils.py (hunk 1 above)
Comment (lunkwill42) [reply src/zino/utils.py:36]: The most popularly used library I've been able to find in a few short minutes of research is https://github.com/saghul/aiodns

Comment (stveit) [reply src/zino/utils.py]: derp, made some changes now

File: src/zino/statemodels.py [hunk 2]
```diff
@@ -350,6 +350,7 @@ class BFDEvent(Event):
     bfdix: Optional[int] = None
     bfddiscr: Optional[int] = None
     bfdaddr: Optional[IPAddress] = None
+    neigh_rdns: Optional[str] = None
```
Comment (johannaengland) [inline src/zino/statemodels.py:353]: According to @runborg  the name needs to be `Neigh-rDNS` ref: https://github.com/Uninett/zino/issues/199#issuecomment-2186359376
```suggestion
    Neigh-rDNS: Optional[str] = None
```

**Review by johannaengland — DISMISSED** [review]

File: src/zino/statemodels.py (hunk 2 above)
Comment (lunkwill42) [reply src/zino/statemodels.py:353]: Indeed, except this is an invalid name in Python, so this doesn't work.  We may have to look into Pydantic field aliases.  One thing is supporting them on deserialization, but I can't recall what we do for serializing to the API, I'll have to check

**Comment by runborg** [comment]
Looking at this there could potentially be multiple dns resolution requests in short time, i think this could be speeded up by using some kind of caching. eg. `lru_cache` or `ttl_cache`

**Comment by stveit** [comment]
> Looking at this there could potentially be multiple dns resolution requests in short time, i think this could be speeded up by using some kind of caching. eg. `lru_cache` or `ttl_cache`

Caching can be added. Should it then have a time-to-live, a certain duration where each result will be deleted after existing for a certain period of time?

**Comment by stveit** [comment]
> Looking at this there could potentially be multiple dns resolution requests in short time, i think this could be speeded up by using some kind of caching. eg. `lru_cache` or `ttl_cache`

Caching can be added. How long do you imagine the TTL should be? and any ideas for maxsize to use?

File: src/zino/statemodels.py (hunk 2 above)
Comment (stveit) [reply src/zino/statemodels.py:353]: This is the case for several values (`alarm_count` and `alarm_type` in AlarmEvent for instance). A simple solution would be to "zinofy" the attribute name in `model_dump_simple_attrs` by replacing `_` with `-`

**Comment by lunkwill42** [comment]
> Caching can be added. How long do you imagine the TTL should be? and any ideas for maxsize to use?

I can't see that Zino 1 does any caching here, so I'm not sure we should bother with this for Zino 2.0.  Post-zino-2 would be fine.

File: src/zino/statemodels.py (hunk 2 above)
Comment (lunkwill42) [reply src/zino/statemodels.py:353]: True, `model_dump_simple_attrs` exists specifically for the legacy protocol - so it should indeed replace `_` with `-` in the output (are there any attrs in the protocol that use an underscore?)

However, that still isn't enough if the *casing* is also important for the protocol clients, which makes me think a Field alias might be the way to go.  `model_dump_simple_attrs` could look up field aliases and use them if they exist, in addition to replacing underscores.

Comment (lunkwill42) [reply src/zino/statemodels.py:353]: On the other hand, question for @runborg : I cannot see that `Neigh-rDNS` is referred to anywhere else in the Zino Tcl code, not even in the Ritz tree (even when searching case insensitively).  Which again leads me to wonder: Does the casing really matter to anybody?

**Comment by stveit** [comment]
> > Caching can be added. How long do you imagine the TTL should be? and any ideas for maxsize to use?
> 
> I can't see that Zino 1 does any caching here, so I'm not sure we should bother with this for Zino 2.0. Post-zino-2 would be fine.

:+1:

File: src/zino/statemodels.py (hunk 2 above)
Comment (stveit) [reply src/zino/statemodels.py:353]: Casing is a good point, if that turns out to matter then something per-field is prob required.

I'd say this outside the scope of this PR anyways, would be weird to introduce special handling of event attributes for neigh_rdns only

**Review by lunkwill42 — APPROVED** [review]
I made a separate issue for the API output stuff discussed in the comments :)

**Pull request merged**